python/ray/serialization.py

```diff
@@ -177,18 +177,19 @@ def object_id_serializer(obj):
 
             def object_id_deserializer(serialized_obj):
                 obj_id, owner_id, owner_address = pickle.loads(serialized_obj)
-                # Must deserialize the object in the core worker before we
-                # create the ObjectID to ensure that the reference is added
-                # before we increment its count to 1.
+                # NOTE(swang): Must deserialize the object first before asking
+                # the core worker to resolve the value. This is to make sure
+                # that the ref count for the ObjectID is greater than 0 by the
+                # time the core worker resolves the value of the object.
+                deserialized_object_id = obj_id[0](obj_id[1][0])
```

Comment:
This fixes a subtle ordering issue where an object is deserialized twice in the same worker. Previously, when we deserialize an object we (1) do object owner registration and add it to the ref counter with num_refs=0, then (2) create the object id. This simplifies it to (1) create the obj id, which creates it with ref 1, then (2) do registration later.